### PR TITLE
[FSDP2] Added pre/post-backward

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118584
* #118298
* #118755
* #118223
* #118136
* #118118
* #118017
* __->__ #118004
* #117975
* #117973
* #117955
* #117950

This PR adds the pre- and post-backward logic:
- **Pre-backward hook:** `FSDPState` and `FSDPParamGroup` define this, and `FSDPState` is responsible for registering since its pre-backward should run even if the `FSDPState` does not manage any parameters (in case it is the root).
- **Post-backward hook:** Only `FSDParamGroup` defines this since the post-backward hook reshards parameters and reduce-scatters gradients (functionality only needed with managed parameters). The `FSDPParamGroup` is responsible for registering this.
- **Post-backward final callback:** `FSDPState` defines this, and each `FSDPParamGroup` defines a `finalize_backward()` to call in the final callback.

### Pre-Backward

The pre-backward hook is registered on the module outputs (that require gradient), and it should run when the first such output has its gradient computed. The hook may run multiple times per backward, once per module forward. Specifically, there will be one `(pre-backward, post-backward)` interval for each of the module's `forward()` calls. This is contrast with the existing FSDP semantics, which only defines a single `(pre-backward, post-backward)` interval that is equivalent to the union of this FSDP's `(pre-backward, post-backward)` intervals. This avoids spiking memory from having multiple modules not resharding and avoids some autograd edge cases.

We implement the pre-backward hook by having a flag that is set upon the 1st calls to disable subsequent calls. This flag could be maintained by FSDP, but for a cleaner design, we augment `register_multi_grad_hook` with a `mode="any"` option and use that instead.

### Post-Backward

The post-backward hook is equivalent to a module full backward hook (`nn.Module.register_full_backward_hook`) except it adds pytree logic to work with data structures other than just flat `Tensor` args passed to `nn.Module.forward`. If we were to use `register_full_backward_hook`, then the hook could fire early (before all gradients for the module have been computed). Most internal models use custom data structures as `forward` inputs, and they find that unifying under pytree is an acceptable solution.

Unlike existing FSDP, we are able to reshard the parameters in the post-backward hook _before_ 'concatenating' the autograd-computed gradients, achieving a lower peak memory usage. (Existing FSDP has `SplitWithSizesBackward` that calls a `CatArrayBatched`, and here we have the reduce-scatter copy-in.)

### Final Callback
The final callback runs as a queued callback to the autograd engine, meaning that it runs at the end of backward.

In the future, if we do not want to wait for the reduce-scatter (or similar for CPU offloading), we can augment the final callback. The code is written such that each reduce-scatter can be waited on separately (via CUDA event).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 1]
```diff
@@ -220,3 +305,20 @@ def _get_param_module_infos(
     if len(param_to_module_info) != len(params):
         raise AssertionError(f"Some parameters are not in the module tree of {module}")
     return [param_to_module_info[param] for param in params]
+
+
+class RegisterPostBackwardHook(torch.autograd.Function):
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:310]: this part feels mysterious to me. is there some api doc we can refer to regarding inheriting ``torch.autograd.Function`` and call ``.apply(tensor)`` ?

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:310]: https://pytorch.org/docs/stable/notes/extending.html#extending-autograd
https://pytorch.org/tutorials/beginner/examples_autograd/two_layer_net_custom_function.html

**Review by weifengpy — APPROVED** [review]

**Comment by awgu** [comment]
I will fix the failing test in the next time I push. It is from rebasing.

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 2]
```diff
@@ -23,18 +36,22 @@ class FSDPState(_State):
...
+        self._is_root: Optional[bool] = None  # root responsible for lazy init
+        self._state_ctx = FSDPStateContext()
+        self._comm_ctx = FSDPCommContext()
         self._training_state: TrainingState = TrainingState.IDLE
         self._pre_forward_hook_handle: Optional[RemovableHandle] = None
+        self._pre_backward_hook_handles: List[RemovableHandle] = []
 
         # Attributes only used on the root state:
-        self._all_states: List[FSDPState] = []
+        self._root_post_backward_final_callback_queued: Optional[bool] = None
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:54]: deleting this in the next push

**Comment by awgu** [comment]
Lint is fixed locally -- will update in next push.

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 3]
```diff
@@ -2,59 +2,74 @@
...
-    # For overlapping current copy-out and next all-gather in forward
-    _all_gather_state: AllGatherStateHolder
 
     def __init__(self):
         super().__init__()
         self._fsdp_param_group: Optional[FSDPParamGroup] = None
-        self._is_root: Optional[bool] = None
+        self._is_root: Optional[bool] = None  # root set during lazy init
+        self._state_ctx = FSDPStateContext()
+        self._comm_ctx = FSDPCommContext()
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:44]: Nice refactor! FSDPState looks much more clean than before!

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 4]
```diff
@@ -2,59 +2,74 @@
...
         self, module: nn.Module, args: Tuple[Any, ...], kwargs: Dict[str, Any]
     ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
         self._lazy_init()
-        if not self._is_root:
+        if self._state_ctx.iter_forward_root is not None:
             return args, kwargs
+        self._state_ctx.iter_forward_root = self
         with torch.profiler.record_function("FSDP::root_pre_forward"):
+            # Wait for optimizer before implicitly prefetched all-gathers
+            current_stream = torch.cuda.current_stream()
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:70]: nit: get the default stream from comm_context?

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 5]
```diff
@@ -2,59 +2,74 @@
...
+class FSDPStateContext:
+    """This has state shared across FSDP states."""
+
+    def __init__(self):
+        # All FSDP states in the root state's module tree
+        self.all_states: List[FSDPState] = []
+        # Iteration's forward root runs the once-per-forward logic; this root
+        # may not be the overall root set by lazy initialization in cases where
+        # only a submodule runs forward
+        self.iter_forward_root: Optional[FSDPState] = None
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:30]: hmmm not fully get this, why this `iter_forward_root` could be different from the root after lazy init?

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 6]
```diff
@@ -124,11 +142,53 @@ def _pre_forward(
...
         if self._fsdp_param_group:
             output = self._fsdp_param_group.post_forward(module, input, output)
+        output = self._register_pre_backward_hook(output)
         self._training_state = TrainingState.IDLE
-        if self._is_root and (all_gather_state := self._all_gather_state.pop()):
-            self._all_gather_copy_in_stream.wait_event(all_gather_state.event)
-            self._all_gather_stream.wait_event(all_gather_state.event)
-            del all_gather_state  # free
+        if self._state_ctx.iter_forward_root is self:
+            if all_gather_state := self._comm_ctx.all_gather_state:
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:148]: add some doc about why here the copy in stream need to wait for the allgather event?

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 7]
```diff
@@ -166,13 +234,46 @@ def _to_unsharded(self):
...
             self._sharded_state = ShardedState.UNSHARDED
 
+    # Hook Registration #
+    def _register_post_backward_hook(
+        self, args: Tuple[Any, ...], kwargs: Dict[str, Any]
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        if not torch.is_grad_enabled():
+            return args, kwargs
+        args_list, args_spec = tree_flatten(args)
+        kwargs_list, kwargs_spec = tree_flatten(kwargs)
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:244]: so it seems like every pre-forward we'll need to run the module inputs post backward grad autograd function and do tree flatten/unflatten, do we worry about CPU overhead here or because it's mostly module level the overhead is tolerable?

**Review by wanchaol — APPROVED** [review]
The logic is actually quite readable, thanks for rafactoring!

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 8]
```diff
@@ -166,13 +234,46 @@ def _to_unsharded(self):
...
+
     # Properties #
     @property
     def _all_gather_process_group(self) -> dist.ProcessGroup:
         mesh_info = self.mesh_info
         assert isinstance(mesh_info, FSDPMeshInfo)
         return mesh_info.shard_process_group
 
+    @property
+    def _reduce_scatter_process_group(self) -> dist.ProcessGroup:
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:272]: Good question! In the next PR, when we add `reshard_after_forward`, the all-gather process group can be a smaller PG if we resharded to that smaller PG after forward. (Because I was pulling code from the mega PR with `reshard_after_forward` implemented already, I directly added this as 2 separate properties.)

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:244]: We do worry about CPU overhead, but at least the overhead is proportional to the number of args/kwargs, which is typically not that high. I am not sure there is a way to avoid this overhead because the tensors here are created newly each iteration.

The main place we got burned on CPU overhead was when there was `O(NP)` overhead for world size `N` and `P` `nn.Parameter`s. I think this overhead here should be far less, but we can see as we run more real jobs.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 9]
```diff
@@ -219,3 +318,16 @@ def _get_param_module_infos(
     if len(param_to_module_info) != len(params):
         raise AssertionError(f"Some parameters are not in the module tree of {module}")
     return [param_to_module_info[param] for param in params]
+
+
+class RegisterPostBackwardHook(torch.autograd.Function):
```
[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:323]: Renamed to `RegisterPostBackwardFunction`

File: torch/distributed/_composable/fsdp/_fsdp_state.py (hunk 6 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:148]: I added a note in `_fsdp_param_group.py` to explain how we overlap copy-in and all-gather, and I added a reference to that note here.
